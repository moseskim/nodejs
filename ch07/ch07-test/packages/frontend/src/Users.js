import './App.css';
import { useUsers } from './Users.hooks'

function User({ name }) {
  return <li style={{ padding: '10px' }}>{name}</li>;
}

// 사용자 정보 표시를 담당하는 컴포넌트
function Users() {
  const { users, setInputText, submit, addCounter } = useUsers();

  const userList = users.map((user) => {
    return <User key={user} name={user} />;
  });

  const handleSubmit = (event) => {
    event.preventDefault();
    submit();
  };

  const handleChange = (event) => {
    setInputText(event.target.value);
  };

  return (
    <div className="App">
      <ul>{userList}</ul>
      <form onSubmit={handleSubmit}>
        <input type="text" onChange={handleChange} />
        <button type="submit">추가</button>
      </form>
      <button onClick={() => addCounter()}>업데이트</button>
    </div>
  );
}

export default Users;
