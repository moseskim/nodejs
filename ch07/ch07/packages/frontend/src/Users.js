import { useState, useEffect } from 'react';
import './App.css';

function User({ name }) {
  return <li style={{ padding: '8px' }}>{name}</li>;
}

const getUsers = async () => {
  const response = await fetch('/api/users');
  const body = response.json();
  return body;
};

function Users() {
  const [users, setUsers] = useState([]);
  const [inputText, setInputText] = useState('');
  const [counter, setCounter] = useState(0);

  useEffect(() => {
    getUsers()
      .then((data) => {
        const users = data.users.map((user) => user.name);
        return users;
      })
      .then((users) => setUsers(users))
      .catch((error) => console.error(error));
  }, [counter]);

  const userList = users.map((user) => {
    return <User key={user} name={user} />;
  });

  const handleSubmit = (event) => {
    event.preventDefault();
    const newUsers = [...users, inputText];
    setUsers(newUsers);
  };

  const handleChange = (event) => {
    setInputText(event.target.value);
  };

  return (
    <div className="App">
      <ul>{userList}</ul>
      <form onSubmit={handleSubmit}>
        <input type="text" onChange={handleChange} />
        <button type="submit">추가</button>
      </form>
      <button onClick={() => setCounter(counter + 1)}>업데이트</button>
    </div>
  );
}

export default Users;
