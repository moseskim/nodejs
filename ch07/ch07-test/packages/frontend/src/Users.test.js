import { render, screen, fireEvent } from '@testing-library/react';
import Users from './Users';
import { useUsers } from './Users.hooks'
import renderer from 'react-test-renderer';

jest.mock('./Users.hooks', () => {
  return {
    useUsers: jest.fn()
  };
});

test('renders Users', () => {
  useUsers.mockImplementation(() => {
    return {
      users: ['alpha', 'bravo']
    }
  });

  render(<Users />);

  expect(screen.getByText('alpha')).toBeInTheDocument();
  expect(screen.getByText('bravo')).toBeInTheDocument();
});

test('추가 버튼의 submit에서 submit이 호출된다', () => {
  const submitMock = jest.fn()

  useUsers.mockImplementation(() => {
    return {
      users: ['alpha', 'bravo'],
      submit: submitMock
    }
  });

  render(<Users />);

  fireEvent.submit(screen.getByText('추가'));

  expect(submitMock).toHaveBeenCalledTimes(1);
});

test('Users 컴포넌트의 스냅샷 테스트', () => {
  useUsers.mockImplementation(() => {
    return {
      users: ['alpha', 'bravo']
    }
  });

  const component = renderer.create(<Users />);

  const tree = component.toJSON();
  // 스냅샷 파일과 비교
  expect(tree).toMatchSnapshot();
});
