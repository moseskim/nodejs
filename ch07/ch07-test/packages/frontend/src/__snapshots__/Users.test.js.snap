// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Users 컴포넌트의 스냅샷 테스트 1`] = `
<div
  className="App"
>
  <ul>
    <li
      style={
        Object {
          "padding": "10px",
        }
      }
    >
      alpha
    </li>
    <li
      style={
        Object {
          "padding": "10px",
        }
      }
    >
      bravo
    </li>
  </ul>
  <form
    onSubmit={[Function]}
  >
    <input
      onChange={[Function]}
      type="text"
    />
    <button
      type="submit"
    >
      추가
    </button>
  </form>
  <button
    onClick={[Function]}
  >
    업데이트
  </button>
</div>
`;
